### DISCUSSION TITLE
Vector Implementation of Multiple TLEs

### DISCUSSION THREAD

**Discussion by cassova** [brandon-rhodes/python-sgp4#89] [Q&A]
I'm looking for a way to pass in an array (like numpy multi-dim array) that contains multiple TLEs for same or diff satellites.  The goal is to take advantage of vector performance to get an `e` array (satellite position) based on a separate array of `jd,fr` (one for each TLE).

It looks like `SatrecArray` only takes in instances of already initialized satellites.  Can this be called without having to create `Satrec` for each satellite?

Thanks,
Nicholas

**Comment by brandon-rhodes** [comment]
I am not aware of any way except to create a `Satrec` for each satellite ­— that is the only class which is used to represent a satellite; there is not an alternative.

Are you finding that the expense of creating one `Satrec` for each satellite is too slow, or is using too much memory?

**Reply by cassova** [reply]
It might work for me as is.  For now, this is what I'm doing using a pandas dataframe:

```python
from sgp4.api import Satrec
from sgp4.conveniences import jday_datetime
from datetime import datetime, timezone

# Create the satellite object (used to find satellite position)
dfs['satobj'] = dfs.apply(lambda x: Satrec.twoline2rv(x['TLE_LINE1'], x['TLE_LINE2']), axis=1)

# Get the Julian date of the EPOCH
dfs.reset_index(inplace=True)
dfs['epoch_julian'] = dfs['EPOCH'].apply(lambda x: jday_datetime(x.replace(tzinfo=timezone.utc)))

# Get the days since 1949 December 31 00:00 UT
# This will be used when creating satobj for the test set
# (this is needed to get the satellite position from generate TLEs)
ref_date = datetime.strptime('12/31/1949 00:00:00', '%m/%d/%Y %H:%M:%S')
dfs['epoch_days'] = dfs['EPOCH'].apply(lambda x: (x-ref_date)/np.timedelta64(1, 'D'))
dfs.set_index('EPOCH', inplace=True)

# Get satellite x,y,z positions
dfs['satpos'] = dfs.apply(lambda x: x['satobj'].sgp4(*x['epoch_julian'])[1], axis=1)
```

On a sample set of 241687 unique TLE records, this takes about 14s on my machine.  While this is reasonable, my end goal is to do this on _all_ TLE data (like gp_history all).  I can probably wait a few hours if that's what it's going to take.

**Reply by cassova** [reply]
Checking the memory, it looks like that could be an issue.

```
TLE_LINE1               30452562
TLE_LINE2               30452562
satobj                 245553992
epoch_julian            11600976
epoch_days               1933496
satpos                  13534472
```

**Reply by brandon-rhodes** [reply]
I will be interested to hear how performance changes (if at all) once you have the chance to try out the `SatrecArray`. I'm not sure, alas, that there's any way around the memory usage, as down at the C++ level those satellite structs just have lots of floats stuffed into them.

**Reply by cassova** [reply]
Unfortunately `SatrecArray` applies the same `jd,fr` dates to all satellites in the array.  While I need to apply a unique value to each so I'm not able to use it.

**Comment by brandon-rhodes** [comment]
Happily, the opposite is the case: it applies a different `jd` and `fr` to each satellite. Could you suggest how the docs at the following link might be improved to make that clearer? I'll be happy to revise them:

https://pypi.org/project/sgp4/#array-acceleration
